lighthouse-core/gather/gatherers/start-url.js

```diff
@@ -9,76 +9,90 @@ const Gatherer = require('./gatherer');
 const manifestParser = require('../../lib/manifest-parser');
 
 class StartUrl extends Gatherer {
-  executeFetchRequest(driver, url) {
-    return driver.evaluateAsync(
-      `fetch('${url}')`
-    );
-  }
-
+  /**
+   * Grab the manifest, extract it's start_url, attempt to `fetch()` it while offline
+   * @param {*} options
+   * @return {{statusCode: number, debugString: ?string}}
+   */
   afterPass(options) {
-    return options.driver.goOnline(options)
-      .then(() => options.driver.getAppManifest())
+    const driver = options.driver;
+    return driver.goOnline(options)
+      .then(() => driver.getAppManifest())
+      .then(response => driver.goOffline(options).then(() => response))
       .then(response => response && manifestParser(response.data, response.url, options.url))
       .then(manifest => {
-        if (!manifest || !manifest.value) {
-          const detailedMsg = manifest && manifest.debugString;
-
-          if (detailedMsg) {
-            const debugString = `Error fetching web app manifest: ${detailedMsg}`;
-            return {
-              statusCode: -1,
-              debugString,
-            };
-          } else {
-            const debugString = `No usable web app manifest found on page ${options.url}`;
-            return {
-              statusCode: -1,
-              debugString,
-            };
-          }
+        const {isReadFailure, reason, startUrl} = this._readManifestStartUrl(manifest);
+        if (isReadFailure) {
+          return {statusCode: -1, debugString: reason};
         }
 
-        if (manifest.value.start_url.debugString) {
-          // Even if the start URL had an error, the browser will still supply a fallback URL.
-          // Therefore, we only set the debugString here and continue with the fetch.
-          return {
-            statusCode: -1,
-            debugString: manifest.value.start_url.debugString,
-          };
-        }
+        return this._attemptManifestFetch(options, startUrl);
+      }).catch(() => {
+        return {statusCode: -1, debugString: 'Unable to fetch start URL via service worker'};
+      });
+  }
+
+  /**
+   * Read the parsed manifest and return failure reasons or the startUrl
+   * @param {Manifest} manifest
+   * @return {{isReadFailure: !boolean, reason: ?string, startUrl: ?startUrl}}
```

Comment:
don't need any of the `!` anymore 🎉  @brendankenny are we going with closure or ts style `?`?

Reply:
no more `!`! Assume everything is non-nullable and add `?` (or `|null`) if you really want something to be nullable

Reply:
also I think you want `reason?: string, startUrl?: startUrl` here since they're optional and not null? If you want to get *real* fancy you can do
```
@return {{isReadFailure: true, reason: string}|{isReadFailure: false, startUrl: string}}
```